File: src/rules/noCallbacks.ts
import { HashSet, Option } from "effect"
import * as ts from "typescript"
import { onNode } from "./ruleCheck.js"
import { createRuleMatch } from "./ruleMatch.js"
import { isInAmbientContext } from "./tsNode.js"
import { callSignatureCheck, hasCallSignature, isVoidType } from "./tsType.js"
import { ExampleSnippet, Rule, RuleExample } from "./types.js"
import type { RuleContext, RuleMatch } from "./types.js"

const ruleId = "no-callbacks"

type CallbackStyleDeclaration =
  | ts.FunctionDeclaration
  | ts.FunctionExpression
  | ts.ArrowFunction
  | ts.MethodDeclaration
  | ts.MethodSignature
  | ts.CallSignatureDeclaration
  | ts.FunctionTypeNode

const isCallbackStyleCandidate = (
  node: ts.Node
): node is CallbackStyleDeclaration => {
  const isFunctionOrExpression =
    ts.isFunctionDeclaration(node) || ts.isFunctionExpression(node)
  const isArrowOrMethod =
    ts.isArrowFunction(node) || ts.isMethodDeclaration(node)
  const isSignature =
    ts.isMethodSignature(node) || ts.isCallSignatureDeclaration(node)
  const isFunctionOrArrow = isFunctionOrExpression || isArrowOrMethod
  const isDirectCallbackKind = isFunctionOrArrow || isSignature

  if (!ts.isFunctionTypeNode(node)) {
    return isDirectCallbackKind
  }

  const typeNode = effectiveCallableTypeNode(node)
  const parent = typeNode.parent
  const isValueDeclaration =
    ts.isVariableDeclaration(parent) || ts.isPropertyDeclaration(parent)

  if (isValueDeclaration) {
    const isTypeAnnotation = parent.type === typeNode
    const initializer = Option.fromNullable(parent.initializer)

    const isNotRuntimeFunction = !Option.exists(
      initializer,
      isRuntimeFunctionLike
    )

    return isTypeAnnotation && isNotRuntimeFunction
  }

  const aliasDeclaration = Option.liftPredicate(ts.isTypeAliasDeclaration)(
    parent
  )
  const hasTypeAliasFunctionType = Option.exists(
    aliasDeclaration,
    isTypeOfAlias(typeNode)
  )
  const propertySignature = Option.liftPredicate(ts.isPropertySignature)(parent)
  const hasPropertySignatureFunctionType = Option.exists(
    propertySignature,
    isTypeOfPropertySignature(typeNode)
  )

  return hasTypeAliasFunctionType || hasPropertySignatureFunctionType
}

const transparentTypeNodeKinds = HashSet.make(
  ts.SyntaxKind.ParenthesizedType,
  ts.SyntaxKind.UnionType,
  ts.SyntaxKind.IntersectionType
)

const effectiveCallableTypeNode = (typeNode: ts.TypeNode): ts.TypeNode =>
  HashSet.has(transparentTypeNodeKinds, typeNode.parent.kind)
    ? effectiveCallableTypeNode(typeNode.parent as ts.TypeNode)
    : typeNode

const isRuntimeFunctionLike = (node: ts.Expression): boolean =>
  ts.isFunctionExpression(node) || ts.isArrowFunction(node)

const isTypeOfAlias =
  (typeNode: ts.TypeNode) =>
  (parent: ts.TypeAliasDeclaration): boolean =>
    parent.type === typeNode

const isTypeOfPropertySignature =
  (typeNode: ts.TypeNode) =>
  (parent: ts.PropertySignature): boolean =>
    parent.type === typeNode

const isFunctionArgument =
  (checker: ts.TypeChecker) =>
  (parameter: ts.ParameterDeclaration): boolean => {
    const parameterType = checker.getTypeAtLocation(parameter)
    const parameterHasCallSignature = hasCallSignature(checker)(parameterType)
    const restToken = Option.fromNullable(parameter.dotDotDotToken)

    if (Option.isNone(restToken)) {
      return parameterHasCallSignature
    }

    const indexType = checker.getIndexTypeOfType(
      parameterType,
      ts.IndexKind.Number
    )
    const elementType = Option.fromNullable(indexType)
    const elementHasCallSignature = Option.exists(
      elementType,
      callSignatureCheck(checker)
    )

    return [parameterHasCallSignature, elementHasCallSignature].some(Boolean)
  }

const isCallbackSignature =
  (context: RuleContext) =>
  (declaration: CallbackStyleDeclaration) =>
  (signature: ts.Signature): boolean => {
    const returnType = context.checker.getReturnTypeOfSignature(signature)
    const returnsVoid = isVoidType(returnType)
    const hasFunctionArgument = declaration.parameters.some(
      isFunctionArgument(context.checker)
    )

    return returnsVoid && hasFunctionArgument
  }

const callbackStyleMatches =
  (context: RuleContext) =>
  (declaration: CallbackStyleDeclaration): ReadonlyArray<RuleMatch> => {
    // A declare statement mirrors a third-party API's existing shape; there is no Effect-returning alternative to describe.
    if (isInAmbientContext(declaration)) {
      return []
    }

    const declaredSignature =
      context.checker.getSignatureFromDeclaration(declaration)
    const signature = Option.fromNullable(declaredSignature)
    const isCallback = Option.exists(
      signature,
      isCallbackSignature(context)(declaration)
    )

    return isCallback
      ? [
          createRuleMatch(context)({
            ruleId,
            node: declaration,
            message:
              "Avoid callback-style functions that accept a function argument and return void.",
            hint:
              "Use Effect instead: wrap third-party callback APIs in an Effect, or declare your " +
              "own API as an Effect-returning function from the start. Ambient declarations " +
              "(declare statements) describing a third-party API are permitted."
          })
        ]
      : []
  }

const check = onNode([
  ts.SyntaxKind.FunctionDeclaration,
  ts.SyntaxKind.FunctionExpression,
  ts.SyntaxKind.ArrowFunction,
  ts.SyntaxKind.MethodDeclaration,
  ts.SyntaxKind.MethodSignature,
  ts.SyntaxKind.CallSignature,
  ts.SyntaxKind.FunctionType
])(isCallbackStyleCandidate)(callbackStyleMatches)

const badExample = new ExampleSnippet({
  filePath: "src/events.ts",
  code: `interface Message {
  readonly data: string
}

interface MessageSocket {
  addEventListener(event: "message", handler: (msg: Message) => void): MessageSocket
}

declare const socket: MessageSocket

export const onMessage = (handler: (msg: Message) => void): void => {
  socket.addEventListener("message", handler)
}`
})

const goodOneShot = new ExampleSnippet({
  filePath: "src/events.ts",
  code: `import { Effect } from "effect"

interface Message {
  readonly data: string
}

interface MessageSocket {
  addEventListener(event: "message", handler: (msg: Message) => void): MessageSocket
}

declare const socket: MessageSocket

type MessageListener = (msg: Message) => void

type MessageResume = (effect: Effect.Effect<Message>) => void

const resumeWithMessage =
  (resume: MessageResume): MessageListener =>
  (msg) => {
    const succeeded = Effect.succeed(msg)
    resume(succeeded)
  }

// One-shot: resolves on the first event, then the Effect completes.
export const onMessage = Effect.async<Message>((resume) => {
  socket.addEventListener("message", resumeWithMessage(resume))

  return Effect.void
})`
})

const goodStream = new ExampleSnippet({
  filePath: "src/messages.ts",
  code: `import { Effect, Stream, StreamEmit } from "effect"

interface Message {
  readonly data: string
}

interface MessageSocket {
  addEventListener(event: "message", handler: (msg: Message) => void): MessageSocket
}

declare const socket: MessageSocket

type MessageListener = (msg: Message) => void

type MessageEmit = StreamEmit.Emit<never, never, Message, void>

const emitMessage =
  (emit: MessageEmit): MessageListener =>
  (msg) => {
    void emit.single(msg)
  }

// Streaming: emits every event until the scope is closed.
export const messages = Stream.async<Message>((emit) => {
  socket.addEventListener("message", emitMessage(emit))

  return Effect.void
})`
})

const example = new RuleExample({
  bad: [badExample],
  good: [goodOneShot, goodStream]
})

export const noCallbacks = new Rule({
  id: ruleId,
  description:
    "Disallow callback-style functions returning void in favor of Effect.",
  example,
  check
})

File: internal/rules/no_callbacks/no_callbacks.go
package no_callbacks

import (
	"github.com/andrueandersoncs/better-typescript/internal/rule"
	"github.com/andrueandersoncs/better-typescript/internal/utils"
	"github.com/andrueandersoncs/typescript-go/ast"
	"github.com/andrueandersoncs/typescript-go/checker"
)

var message = rule.RuleMessage{
	Id:          "noCallbacks",
	Description: "Avoid callback-style void APIs.",
	Help:        "Return an Effect from the operation instead of accepting a callback.",
}

func isInAmbientContext(node *ast.Node) bool {
	for current := node; current != nil; current = current.Parent {
		if ast.IsSourceFile(current) {
			return current.AsSourceFile().IsDeclarationFile
		}
		if current.ModifierFlags()&ast.ModifierFlagsAmbient != 0 {
			return true
		}
	}
	return false
}

func effectiveCallableTypeNode(node *ast.Node) *ast.Node {
	for node.Parent != nil && (ast.IsParenthesizedTypeNode(node.Parent) || ast.IsUnionTypeNode(node.Parent) || ast.IsIntersectionTypeNode(node.Parent)) {
		node = node.Parent
	}
	return node
}

func isRuntimeFunctionLike(node *ast.Node) bool {
	return node != nil && (ast.IsFunctionExpression(node) || ast.IsArrowFunction(node))
}

func isCallbackStyleCandidate(node *ast.Node) bool {
	if !ast.IsFunctionTypeNode(node) {
		return ast.IsFunctionDeclaration(node) || ast.IsFunctionExpression(node) || ast.IsArrowFunction(node) || ast.IsMethodDeclaration(node) || ast.IsMethodSignatureDeclaration(node) || ast.IsCallSignatureDeclaration(node)
	}
	typeNode := effectiveCallableTypeNode(node)
	parent := typeNode.Parent
	if parent == nil {
		return false
	}
	if ast.IsVariableDeclaration(parent) || ast.IsPropertyDeclaration(parent) {
		return parent.Type() == typeNode && !isRuntimeFunctionLike(parent.Initializer())
	}
	if ast.IsTypeAliasDeclaration(parent) {
		return parent.Type() == typeNode
	}
	if ast.IsPropertySignatureDeclaration(parent) {
		return parent.Type() == typeNode
	}
	return false
}

func hasCallSignature(ctx rule.RuleContext, valueType *checker.Type) bool {
	return valueType != nil && len(utils.GetCallSignatures(ctx.TypeChecker, valueType)) > 0
}

func parameterIsFunction(ctx rule.RuleContext, parameter *ast.Node) bool {
	parameterType := ctx.TypeChecker.GetTypeAtLocation(parameter)
	if hasCallSignature(ctx, parameterType) {
		return true
	}
	if parameter.AsParameterDeclaration().DotDotDotToken == nil || parameterType == nil || checker.Type_flags(parameterType)&checker.TypeFlagsObject == 0 || checker.Type_objectFlags(parameterType)&checker.ObjectFlagsReference == 0 {
		return false
	}
	arguments := checker.Checker_getTypeArguments(ctx.TypeChecker, parameterType)
	return len(arguments) > 0 && hasCallSignature(ctx, arguments[0])
}

func checkDeclaration(ctx rule.RuleContext, node *ast.Node) {
	if !isCallbackStyleCandidate(node) || isInAmbientContext(node) {
		return
	}
	signature := ctx.TypeChecker.GetSignatureFromDeclaration(node)
	if signature == nil {
		return
	}
	returnType := ctx.TypeChecker.GetReturnTypeOfSignature(signature)
	if returnType == nil || !utils.IsTypeFlagSet(returnType, checker.TypeFlagsVoid) {
		return
	}
	for _, parameter := range node.Parameters() {
		if parameterIsFunction(ctx, parameter) {
			ctx.ReportNode(node, message)
			return
		}
	}
}

func listener(ctx rule.RuleContext) func(*ast.Node) {
	return func(node *ast.Node) { checkDeclaration(ctx, node) }
}

var NoCallbacksRule = rule.Rule{
	Name: "no-callbacks",
	Run: func(ctx rule.RuleContext, options any) rule.RuleListeners {
		check := listener(ctx)
		return rule.RuleListeners{
			ast.KindFunctionDeclaration: check,
			ast.KindFunctionExpression:  check,
			ast.KindArrowFunction:       check,
			ast.KindMethodDeclaration:   check,
			ast.KindMethodSignature:     check,
			ast.KindCallSignature:       check,
			ast.KindFunctionType:        check,
		}
	},
}
